pkg/i2gw/providers/openapi3/openapi.go

```diff
@@ -0,0 +1,111 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi3
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
+)
+
+const (
+	// The ProviderName returned to the provider's registry.
+	ProviderName = "openapi3"
+
+	BackendFlag      = "backend"
+	GatewayClassFlag = "gateway-class-name"
+	TlsSecretFlag    = "gateway-tls-secret"
+)
+
+func init() {
+	i2gw.ProviderConstructorByName[ProviderName] = NewProvider
+
+	i2gw.RegisterProviderSpecificConf(ProviderName, i2gw.ProviderSpecificConf{
+		Name:        BackendFlag,
+		Description: "The name of the backend service to use in the HTTPRoutes",
+	})
+
+	i2gw.RegisterProviderSpecificConf(ProviderName, i2gw.ProviderSpecificConf{
+		Name:        GatewayClassFlag,
+		Description: "The name of the gateway class to use in the Gateways",
+	})
+
+	i2gw.RegisterProviderSpecificConf(ProviderName, i2gw.ProviderSpecificConf{
+		Name:        TlsSecretFlag,
+		Description: "The name of the secret for the TLS certificate references in the Gateways",
+	})
+}
+
+type Provider struct {
+	storage   Storage
+	converter Converter
+}
+
+var _ i2gw.Provider = &Provider{}
+
+// NewProvider returns an implementation of i2gw.Provider that converts OpenAPI specs to Gateway API resources.
+func NewProvider(conf *i2gw.ProviderConf) i2gw.Provider {
+	return &Provider{
+		storage:   NewResourceStorage(),
+		converter: NewConverter(conf),
+	}
+}
+
+// ReadResourcesFromCluster reads OpenAPI specs stored in the Kubernetes cluster. UNIMPLEMENTED.
+func (p *Provider) ReadResourcesFromCluster(_ context.Context) error {
+	return nil
+}
+
+// ReadResourcesFromFile reads OpenAPI specs from a JSON or YAML file.
+func (p *Provider) ReadResourcesFromFile(ctx context.Context, filename string) error {
+	spec, err := readSpecFromFile(ctx, filename)
+	if err != nil {
+		return fmt.Errorf("failed to read resources from file: %w", err)
+	}
+
+	p.storage.Clear()
+	if spec != nil {
+		p.storage.AddResource(spec)
+	}
+
+	return nil
+}
+
+// ToGatewayAPI converts stored OpenAPI specs to Gateway API resources.
+func (p *Provider) ToGatewayAPI() (i2gw.GatewayResources, field.ErrorList) {
+	return p.converter.Convert(p.storage)
+}
+
+func readSpecFromFile(ctx context.Context, filename string) (*openapi3.T, error) {
+	loader := openapi3.NewLoader()
+	spec, err := loader.LoadFromFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load OpenAPI spec: %w", err)
+	}
+
+	if err := spec.Validate(ctx); err != nil {
+		log.Printf("%s provider: invalid OpenAPI 3.x spec: %v", ProviderName, err)
+		return nil, nil
```

Comment:
why nil and not error?

Reply:
What I meant in https://github.com/kubernetes-sigs/ingress2gateway/pull/157#discussion_r1590001944 is the latter. And yes, for that comment it was important that `ReadResourcesFromFile` would not return error, otherwise the tool will fail for all `providers`. 

However, I opened #158 to change the default mode to choose no providers. Will soon get to do this. 

So I think for now you should make `readSpecFromFile` return an error, log this error in `ReadResourcesFromFile` and return nil. Once I make a PR for #158 I will change that to return the error.

Reply:
Did you implemented this?

Reply:
I guess I missed this one. Doing it now...

Reply:
Done

Reply:
Oops. Forgot to fix the tests. Doing it now.

Reply:
Now done.